tests/lib/ifaces/linux_bridge_iface_test.py

```diff
@@ -0,0 +1,420 @@
+
+        des_iface = LinuxBridgeIface(des_iface_info)
+        cur_iface = LinuxBridgeIface(cur_iface_info)
+        des_iface.merge(cur_iface)
+
+        assert des_iface.to_dict() == expected_iface_info
+
+    def test_desire_port_name_partial_merge_from_current_with_vlan_filter(
+        self,
+    ):
+        cur_iface_info = gen_bridge_iface_info_with_vlan_filter()
+        des_iface_info = gen_bridge_iface_info_with_vlan_filter()
+        expected_port_config = deepcopy(SLAVE2_PORT_CONFIG)
+        expected_port_config[Port.VLAN_SUBTREE] = SLAVE2_VLAN_CONFIG_TRUNK
+        expected_iface_info = gen_bridge_iface_info()
+        expected_iface_info[LB.CONFIG_SUBTREE][LB.PORT_SUBTREE] = [
+            expected_port_config
+        ]
+        des_iface_info[LB.CONFIG_SUBTREE][LB.PORT_SUBTREE] = [
+            {Port.NAME: SLAVE2_IFACE_NAME},
+        ]
+
+        des_iface = LinuxBridgeIface(des_iface_info)
+        cur_iface = LinuxBridgeIface(cur_iface_info)
+        des_iface.merge(cur_iface)
+
+        assert des_iface.to_dict() == expected_iface_info
+
+    def test_fix_access_mode_with_enable_native(self):
+        cur_iface_info = gen_bridge_iface_info_with_vlan_filter()
+        des_iface_info = gen_bridge_iface_info_with_vlan_filter()
+        expected_iface_info = gen_bridge_iface_info_with_vlan_filter()
+        cur_vlan_conf = cur_iface_info[LB.CONFIG_SUBTREE][LB.PORT_SUBTREE][0][
+            Port.VLAN_SUBTREE
+        ]
+        cur_vlan_conf[Vlan.ENABLE_NATIVE] = True
+        cur_vlan_conf[Vlan.MODE] = Vlan.Mode.TRUNK
+
+        des_iface = LinuxBridgeIface(des_iface_info)
+        cur_iface = LinuxBridgeIface(cur_iface_info)
+        des_iface.merge(cur_iface)
+        des_iface.pre_edit_validation_and_cleanup()
+
+        assert des_iface.to_dict() == expected_iface_info
+
+    def test_merge_fix_tag_with_trunk_disabled_native(self):
+        cur_iface_info = gen_bridge_iface_info_with_vlan_filter()
+        des_iface_info = gen_bridge_iface_info_with_vlan_filter()
+        des_iface_info[LB.CONFIG_SUBTREE][LB.PORT_SUBTREE][0][
+            Port.VLAN_SUBTREE
+        ] = {
+            Vlan.ENABLE_NATIVE: False,
+            Vlan.MODE: Vlan.Mode.TRUNK,
+            Vlan.TRUNK_TAGS: [{Vlan.TrunkTags.ID: 101}],
+        }
+        expected_iface_info = deepcopy(des_iface_info)
+
+        des_iface = LinuxBridgeIface(des_iface_info)
+        cur_iface = LinuxBridgeIface(cur_iface_info)
+        des_iface.merge(cur_iface)
+        des_iface.pre_edit_validation_and_cleanup()
+
+        assert des_iface.to_dict() == expected_iface_info
+
+    def test_remove_trunk_tag_if_access_mode(self):
+        cur_iface_info = gen_bridge_iface_info_with_vlan_filter()
+        des_iface_info = gen_bridge_iface_info_with_vlan_filter()
+        des_iface_info[LB.CONFIG_SUBTREE][LB.PORT_SUBTREE][1][
+            Port.VLAN_SUBTREE
+        ] = {
+            Vlan.MODE: Vlan.Mode.ACCESS,
```

Comment:
This actually leads to an interesting question: would this be an allowed 'update' ? (e.g. if the current state was a trunk port)

Wouldn't you need to also specify the `tag` in this scenario ?.. What if the `current state` is a trunk port w/ native vlan enabled & configured ?...

Reply:
> This actually leads to an interesting question: would this be an allowed 'update' ? (e.g. if the current state was a trunk port)

The design of nmstate merging process is:
 * When user define invalid/conflicting state, we raise Exception.
 * When merging cause invalid/conflicting state, we fix it internally.

For this test case, the merge process should remove the `Vlan.TRUNK_TAGS` from current
when user is switching that port from trunk to access.

> 
> Wouldn't you need to also specify the `tag` in this scenario ?.. What if the `current state` is a trunk port w/ native vlan enabled & configured ?...

For any config still valid after mode change, they will be preserved. If you think that might
confusing people, I can do the same thing bond mode changing did: When mode change, whole configure is required, no merging.
But this should be done in separate PR, let me know if you want it.

Reply:
> > This actually leads to an interesting question: would this be an allowed 'update' ? (e.g. if the current state was a trunk port)
> 
> The design of nmstate merging process is:
> 
> * When user define invalid/conflicting state, we raise Exception.
> * When merging cause invalid/conflicting state, we fix it internally.
> 
> For this test case, the merge process should remove the `Vlan.TRUNK_TAGS` from current
> when user is switching that port from trunk to access.

I'm just trying to understand what would happen.

> 
> > Wouldn't you need to also specify the `tag` in this scenario ?.. What if the `current state` is a trunk port w/ native vlan enabled & configured ?...
> 
> For any config still valid after mode change, they will be preserved. If you think that might
> confusing people, I can do the same thing bond mode changing did: When mode change, whole configure is required, no merging.
> But this should be done in separate PR, let me know if you want it.

I think it is confusing **without** it. 

If I understand correctly, update from trunk to access port (if you don't specify the tag) would **work** for a native vlan, and **fail** for a regular vlan. I'm not sure that's a good thing (bug vs feature :) )

I would prefer if you change the behavior to the suggested alternative, but let's see what others think.

Reply:
I have created github issue https://github.com/nmstate/nmstate/issues/1080 to track this.

This PR is way too big to add more stuff. The current master branch will merge config even
user switch bond mode, and this PR is present the old behavior.

Reply:
Thanks.